File: franka_simulink/blocks/franka_robot_model_jacobian_sfunction.m
function franka_robot_model_jacobian_sfunction(block)
%FRANKA_ROBOT_MODEL_JACOBIAN Compute Jacobian matrix using libfranka Model
%
%  This block computes the 6x7 Jacobian matrix from explicit inputs using
%  the robot's kinematic/dynamic model. It integrates with FrankaRobotManager
%  to share the robot connection with other blocks.
%
%  Parameters:
%    robot_ip      - IP address of the Franka robot (for model lookup)
%    jacobian_type - 0 = zeroJacobian (base frame), 1 = bodyJacobian (body frame)
%    frame         - Target frame: 0-6 = Joint1-7, 7 = Flange, 8 = EndEffector, 9 = Stiffness
%
%  Inputs:
%    q       - Joint positions [rad] (7x1)
%    F_T_EE  - End effector in flange frame (4x4 column-major as 16x1)
%    EE_T_K  - Stiffness frame in EE frame (4x4 column-major as 16x1)
%
%  Outputs:
%    jacobian - Jacobian matrix (6x7)
%
%  Usage Notes:
%    - For current robot state Jacobians, use the model_data output from
%      franka_robot block (FrankaModelDataBus.jacobian, jacobian_body)
%    - This block is for computing Jacobians from arbitrary configurations
%    - Works both standalone AND under function-called subsystems
%
%  Copyright (c) 2025 Franka Robotics GmbH
%  This file is subject to the terms and conditions defined in the file
%  'LICENSE', which is part of this package

setup(block);

function setup(block)

    % Register parameters: robot_ip, jacobian_type, frame
    block.NumDialogPrms     = 3;
    block.DialogPrmsTunable = {'Nontunable', 'Nontunable', 'Nontunable'};
    
    % Register number of ports
    % Inputs: q(7), F_T_EE(16), EE_T_K(16)
    block.NumInputPorts  = 3;
    block.NumOutputPorts = 1;

    % Setup port properties to be inherited or dynamic
    block.SetPreCompInpPortInfoToDynamic;
    block.SetPreCompOutPortInfoToDynamic;

    % Input 1: q - Joint positions (7x1)
    block.InputPort(1).Dimensions  = 7;
    block.InputPort(1).DatatypeID  = 0;  % double
    block.InputPort(1).Complexity  = 'Real';
    block.InputPort(1).DirectFeedthrough = true;

    % Input 2: F_T_EE - End effector in flange frame (16x1)
    block.InputPort(2).Dimensions  = 16;
    block.InputPort(2).DatatypeID  = 0;  % double
    block.InputPort(2).Complexity  = 'Real';
    block.InputPort(2).DirectFeedthrough = true;

    % Input 3: EE_T_K - Stiffness frame in EE frame (16x1)
    block.InputPort(3).Dimensions  = 16;
    block.InputPort(3).DatatypeID  = 0;  % double
    block.InputPort(3).Complexity  = 'Real';
    block.InputPort(3).DirectFeedthrough = true;

    % Output: jacobian (6x7 column-major)
    block.OutputPort(1).Dimensions  = [6, 7];
    block.OutputPort(1).DatatypeID  = 0;  % double
    block.OutputPort(1).Complexity  = 'Real';
    block.OutputPort(1).SamplingMode = 'Sample';
    
    % Inherited sample time: works in both standalone and function-called contexts
    block.SampleTimes = [-1 0];

    block.SimStateCompliance = 'DefaultSimState';

    % Register methods
    block.RegBlockMethod('CheckParameters',         @CheckPrms);
    block.RegBlockMethod('ProcessParameters',       @ProcessPrms);
    block.RegBlockMethod('PostPropagationSetup',    @DoPostPropSetup);
    block.RegBlockMethod('Start',                   @Start); 
    block.RegBlockMethod('WriteRTW',                @WriteRTW);
    block.RegBlockMethod('Outputs',                 @Outputs);

function CheckPrms(block)
    % Validate robot_ip parameter
    robot_ip = block.DialogPrm(1).Data;
    if ~ischar(robot_ip) && ~isstring(robot_ip)
        error('robot_ip must be a string');
    end
    
    % Validate jacobian_type (0 or 1)
    jacobian_type = block.DialogPrm(2).Data;
    if jacobian_type ~= 0 && jacobian_type ~= 1
        error('jacobian_type must be 0 (zero) or 1 (body)');
    end
    
    % Validate frame (0-9)
    frame = block.DialogPrm(3).Data;
    if frame < 0 || frame > 9
        error('frame must be 0-9');
    end

function DoPostPropSetup(block)
    % No DWork needed

function ProcessPrms(block)
    block.AutoUpdateRuntimePrms;

function Start(block)
    % Initialization - nothing to do here

function Outputs(block)
    % For simulation only - actual implementation is in TLC
    block.OutputPort(1).Data = zeros(6, 7);

function WriteRTW(block)
    % Write parameters to RTW file for TLC access
    robot_ip = block.DialogPrm(1).Data;
    jacobian_type = block.DialogPrm(2).Data;
    frame = block.DialogPrm(3).Data;
    
    robot_ip = char(['''', robot_ip, '''']);
    
    block.WriteRTWParam('string', 'robot_ip', robot_ip);
    block.WriteRTWParam('matrix', 'jacobian_type', int8(jacobian_type));
    block.WriteRTWParam('matrix', 'frame', int8(frame));

File: franka_simulink/blocks/franka_robot_model_jacobian_sfunction.cpp
/*
 * franka_robot_model_jacobian_sfunction.cpp - C++ Level-2 S-Function
 *
 * Computes the 6x7 Jacobian matrix using libfranka's Model class.
 *
 * Parameters:
 *   0. jacobian_type - 0 = zeroJacobian (base frame), 1 = bodyJacobian
 *   1. frame         - 0-6 = Joint1-7, 7 = Flange, 8 = EndEffector, 9 = Stiffness
 *
 * Inputs:
 *   0. robot_ip (16x1 uint8)  - Robot IP as ASCII chars
 *   1. q        (7x1 double)  - Joint positions [rad]
 *   2. F_T_EE   (4x4 double)  - End effector in flange frame
 *   3. EE_T_K   (4x4 double)  - Stiffness frame in EE frame
 *
 * Outputs:
 *   0. jacobian (6x7 double)  - Jacobian matrix
 *
 * Copyright (c) 2025 Franka Robotics GmbH
 */

#define S_FUNCTION_NAME  franka_robot_model_jacobian_sfunction
#define S_FUNCTION_LEVEL 2

#include "simstruc.h"

/* Parameters */
#define PARAM_JACOBIAN_TYPE  0
#define PARAM_FRAME          1
#define NUM_PARAMS           2

/* Input port indices */
#define IN_ROBOT_IP   0
#define IN_Q          1
#define IN_F_T_EE     2
#define IN_EE_T_K     3
#define NUM_INPUTS    4

/* Output port indices */
#define OUT_JACOBIAN  0
#define NUM_OUTPUTS   1

/* ========================================================================
 * mdlInitializeSizes - Initialize block sizes
 * ======================================================================== */
static void mdlInitializeSizes(SimStruct *S)
{
    ssSetNumSFcnParams(S, NUM_PARAMS);
    
#if defined(MATLAB_MEX_FILE)
    if (ssGetNumSFcnParams(S) != ssGetSFcnParamsCount(S)) {
        return;
    }
#endif

    /* Make parameters non-tunable */
    ssSetSFcnParamTunable(S, PARAM_JACOBIAN_TYPE, SS_PRM_NOT_TUNABLE);
    ssSetSFcnParamTunable(S, PARAM_FRAME, SS_PRM_NOT_TUNABLE);

    ssSetNumContStates(S, 0);
    ssSetNumDiscStates(S, 0);
    
    /* INPUT PORTS */
    if (!ssSetNumInputPorts(S, NUM_INPUTS)) return;
    
    /* Port 0: robot_ip (16 ASCII chars) */
    ssSetInputPortWidth(S, IN_ROBOT_IP, 16);
    ssSetInputPortDataType(S, IN_ROBOT_IP, SS_UINT8);
    ssSetInputPortDirectFeedThrough(S, IN_ROBOT_IP, 1);
    ssSetInputPortRequiredContiguous(S, IN_ROBOT_IP, 1);
    
    /* Port 1: q - Joint positions (7x1) */
    ssSetInputPortWidth(S, IN_Q, 7);
    ssSetInputPortDataType(S, IN_Q, SS_DOUBLE);
    ssSetInputPortDirectFeedThrough(S, IN_Q, 1);
    ssSetInputPortRequiredContiguous(S, IN_Q, 1);
    
    /* Port 2: F_T_EE (4x4) */
    ssSetInputPortMatrixDimensions(S, IN_F_T_EE, 4, 4);
    ssSetInputPortDataType(S, IN_F_T_EE, SS_DOUBLE);
    ssSetInputPortDirectFeedThrough(S, IN_F_T_EE, 1);
    ssSetInputPortRequiredContiguous(S, IN_F_T_EE, 1);
    
    /* Port 3: EE_T_K (4x4) */
    ssSetInputPortMatrixDimensions(S, IN_EE_T_K, 4, 4);
    ssSetInputPortDataType(S, IN_EE_T_K, SS_DOUBLE);
    ssSetInputPortDirectFeedThrough(S, IN_EE_T_K, 1);
    ssSetInputPortRequiredContiguous(S, IN_EE_T_K, 1);
    
    /* OUTPUT PORTS */
    if (!ssSetNumOutputPorts(S, NUM_OUTPUTS)) return;
    
    /* Port 0: jacobian (6x7) */
    ssSetOutputPortMatrixDimensions(S, OUT_JACOBIAN, 6, 7);
    ssSetOutputPortDataType(S, OUT_JACOBIAN, SS_DOUBLE);
    
    /* SAMPLE TIME */
    ssSetNumSampleTimes(S, 1);
    
    /* WORK VECTORS */
    ssSetNumRWork(S, 0);
    ssSetNumIWork(S, 0);
    ssSetNumPWork(S, 0);
    ssSetNumModes(S, 0);
    ssSetNumNonsampledZCs(S, 0);
    ssSetNumDWork(S, 0);
    
    /* Options */
    ssSetSimStateCompliance(S, USE_DEFAULT_SIM_STATE);
    ssSetOptions(S, SS_OPTION_CALL_TERMINATE_ON_EXIT);
}

/* ========================================================================
 * mdlInitializeSampleTimes - Initialize sample times
 * ======================================================================== */
static void mdlInitializeSampleTimes(SimStruct *S)
{
    ssSetSampleTime(S, 0, INHERITED_SAMPLE_TIME);
    ssSetOffsetTime(S, 0, 0.0);
}

/* ========================================================================
 * mdlOutputs - Compute outputs (simulation only)
 * ======================================================================== */
static void mdlOutputs(SimStruct *S, int_T tid)
{
    real_T *jacobian = ssGetOutputPortRealSignal(S, OUT_JACOBIAN);
    
    for (int i = 0; i < 42; i++) {
        jacobian[i] = 0.0;
    }
    
    UNUSED_ARG(tid);
}

/* ========================================================================
 * mdlTerminate - Cleanup
 * ======================================================================== */
static void mdlTerminate(SimStruct *S)
{
    UNUSED_ARG(S);
}

/* ========================================================================
 * mdlRTW - Write parameters to RTW file for code generation
 * ======================================================================== */
#if defined(MATLAB_MEX_FILE)
#define MDL_RTW
static void mdlRTW(SimStruct *S)
{
    int jacobian_type = static_cast<int>(mxGetScalar(ssGetSFcnParam(S, PARAM_JACOBIAN_TYPE)));
    int frame = static_cast<int>(mxGetScalar(ssGetSFcnParam(S, PARAM_FRAME)));
    
    if (!ssWriteRTWParamSettings(S, 2,
            SSWRITE_VALUE_NUM, "jacobian_type", (real_T)jacobian_type,
            SSWRITE_VALUE_NUM, "frame", (real_T)frame)) {
        return;
    }
}
#endif

/* ========================================================================
 * Required S-function trailer
 * ======================================================================== */
#ifdef MATLAB_MEX_FILE
#include "simulink.c"
#else
#include "cg_sfun.h"
#endif

